lib/MySQL_Variables.cpp

```diff
@@ -0,0 +1,335 @@
+#include "MySQL_Variables.h"
+#include "proxysql.h"
+
+#include "MySQL_Session.h"
+#include "MySQL_Data_Stream.h"
+#include "SpookyV2.h"
+
+#include <sstream>
+
+extern const MARIADB_CHARSET_INFO * proxysql_find_charset_nr(unsigned int nr);
+extern MARIADB_CHARSET_INFO * proxysql_find_charset_name(const char *name);
+
+MySQL_Variables::MySQL_Variables(MySQL_Session* _session) {
+	assert(_session);
+	session = _session;
+
+	for (auto i = 0; i < SQL_NAME_LAST; i++) {
+		switch(i) {
+		case SQL_SAFE_UPDATES:
+		case SQL_SELECT_LIMIT:
+		case SQL_SQL_MODE:
+		case SQL_TIME_ZONE:
+		case SQL_CHARACTER_SET_RESULTS:
+		case SQL_CHARACTER_SET_CONNECTION:
+		case SQL_CHARACTER_SET_CLIENT:
+		case SQL_CHARACTER_SET_DATABASE:
+		case SQL_ISOLATION_LEVEL:
+		case SQL_TRANSACTION_READ:
+		case SQL_SESSION_TRACK_GTIDS:
+		case SQL_SQL_AUTO_IS_NULL:
+		case SQL_COLLATION_CONNECTION:
+		case SQL_NET_WRITE_TIMEOUT:
+		case SQL_MAX_JOIN_SIZE:
+			updaters[i] = new Generic_Updater();
+			break;
+		case SQL_CHARACTER_SET:
+			updaters[i] = new Charset_Updater();
+			break;
+		case SQL_SET_NAMES:
+			updaters[i] = new Names_Updater();
+			break;
+		default:
+			updaters[i] = NULL;
+		}
+	}
+}
+
+MySQL_Variables::~MySQL_Variables() {
+	for (auto u : updaters)
+		delete u;
+}
+
+void print_backtrace(void);
+
+void MySQL_Variables::client_set_value(int idx, const char* value) {
```

Comment:
This function doesn't check if the client exists.

I see this function called by `MySQL_Connection::set_charset()` , that is in turned called by `MySQL_Connection::connect_start()` .
`MySQL_Connection::connect_start()` creates a connection to the backend.
Some sessions do not have a client connection, for example sessions used by mirroring.
Therefore, trying to set a variable on a not existing client will cause a crash

Reply:
Fixed

Reply:
Fixed